mod.ts

```diff
@@ -0,0 +1,81 @@
+const canElideFrames = 'captureStackTrace' in Error
+const startStackFrames = new WeakMap()
+
+interface Result {
+  name: 'AssertionError' | 'AssertionResult'
+  ok: boolean
+  toJSON(...args: unknown[]): Record<string, unknown>
+}
+
+export class AssertionError<T> extends Error implements Result {
+  [key: string]: unknown
+
+  get name(): 'AssertionError' {
+    return 'AssertionError'
+  }
+
+  get ok() {
+    return false
+  }
+
+  constructor(public message = 'Unspecified AssertionError', props?: T, ssf?: Function) {
+    super(message)
+    if (canElideFrames && ssf) startStackFrames.set(this, ssf)
+    for (const key in props)
+      if (!(key in this))
+        // @ts-ignore
+        this[key] = props[key];
+  }
+
+  get stack() {
+    if (canElideFrames) {
+      return (Error as any).captureStackTrace(this, startStackFrames.get(this) || AssertionError);
+    } else {
+      return super.stack
+    }
+  }
+
+  toJSON(stack: boolean): Record<string, unknown> {
+    return {
+      ...this,
+      message: this.message,
+      ok: false,
+      // include stack if exists and not turned off
+      stack: stack !== false ? this.stack : undefined,
+    }
+  }
+
+}
+
+export class AssertionResult<T> implements Result {
+  [key: string]: unknown
+
+  get name(): 'AssertionResult' {
+    return 'AssertionResult'
+  }
+
+  get ok() {
+    return true
+  }
+
+  constructor(props?: T, ssf?: Function) {
+    for (const key in props)
+      if (!(key in this))
+        // @ts-ignore
+        this[key] = props[key];
```

Comment:
I find it hard to read by just indentation. Could we wrap these in curlies?

```suggestion
    for (const key in props) {
      if (!(key in this)) {
        // @ts-ignore
        this[key] = props[key];
      }
    }
```